File: Laguna-2015(prog-evol-2015).py
# -*- coding: UTF-8 -*-
"""
Autor: Jhordan Silveira de Borba (sbjhordan@gmail.com)
Site: https://sites.google.com/view/jdan/
"""

#Bibliotecas
import numpy as np                                      # Biblioteca de funções matemáticas
import copy                                             # Biblioteca com funções para copiar

#CONDIÇÕES INICIAIS ------------------------------------------------------
maxt = 15000    #tiempo total de cada realizacion
Lx   = 100      #tamaño del sustrato
Ly   = 100

#Fracciones iniciales
D    = 0.3  #fraccion inicial de sitios destruidos
x10  = 0.5  #fraccion inicial de sitios ocupados por x1
x20  = 0.5  #fraccion inicial de sitios ocupados por x2
y0   = 0.5  #fraccion inicial de sitios ocupados por y

#Tasas               
cx1   = 0.05   #colonizaciones       
cx2   = 0.1        
cy    = 0.015     
ex1   = 0.05   #extinciones       
ex2   = 0.01      
ey    = 0.017
mx1y  = 0.2    #depredaciones       
mx2y  = 0.8            
depx1 = ex1+mx1y       #mortalidad aumentada=extincion+depredacion
depx2 = ex2+mx2y


g = open("psd.dat", "r") 

#Inicializaciones
s  = np.full((Lx, Ly), [1])   #matriz de habitat (1=sitio disponible, 0=destruido)
x1 = np.full((Lx, Ly), [0])   #1 = sitio ocupado, 0 sitio vazio
x2 = np.full((Lx, Ly), [0])   #1 = sitio ocupado, 0 sitio vazio
y  = np.full((Lx, Ly), [0])   #1 = sitio ocupado, 0 sitio vazio

for i in range(1,Lx-1): 
  for j in range (1,Ly-1): 
    rnd=float(g.readline())#np.random.rand()
    if (rnd < D): s[i,j]=0
    rnd=float(g.readline())#np.random.rand()
    if (rnd < x10 and s[i,j]==1): x1[i,j]=1
    rnd=float(g.readline())#np.random.rand()
    if (rnd < x20 and s[i,j]==1): x2[i,j]=1
    rnd=float(g.readline())#np.random.rand()
    if ((x1[i,j]==1 or x2[i,j]==1) and rnd < y0): y[i,j]=1
    
for i in range(Lx): #destruimos los bordes del habitat
  s[i,0]=0
  s[i,Ly-1]=0
for j in range(Ly):
  s[0,j]=0
  s[Lx-1,j]=0

#-------------------------------------------------------
#Lazo temporal
fx1=sum(sum(x1))/((Lx-1)*(Ly-1)) #Original -1, acredito que deve ser -2
fx2=sum(sum(x2))/((Lx-1)*(Ly-1))
fy =sum(sum(y)) /((Lx-1)*(Ly-1))
f = open("minha.dat", "w")
f.write("	"+str(0)+"	"+str(fx1)+"	"+str(fx2)+"	"+str(fy)+"\n")

for it in range(maxt):
  if (float(it)%(float(maxt)/100)==0.):
    print("Porcentagem: "+str(it*100/maxt))  # Exibe o passo atual

  x1old=copy.copy(x1)              #poblaciones del paso anterior
  x2old=copy.copy(x2)
  yold=copy.copy(y)
  
  #Recorro la malla 
  for i in range (1,Lx-1):
    for j in range (1,Ly-1):
        
      #colonizaciones
      if (x1old[i,j]==0 and s[i,j]==1):                           #colonizacion de x1 si el sitio esta libre de x1 y no destruido:
        nvec=x1old[i-1,j]+x1old[i+1,j]+x1old[i,j-1]++x1old[i,j+1] #cantidad de vecinos ocupados
        p=1.0-(1.0-cx1)**nvec                                     #probabilidad de ser colonizado= 1 - prob de quedar libre
        rnd=float(g.readline())#np.random.rand() 
        if (rnd < p):x1[i,j]=1
      
      if (x2old[i,j]==0 and s[i,j]==1 and x1old[i,j]==0):         #colonizacion de x2 si el sitio esta libre de x1 y x2, y no destruido: 
        nvec=x2old[i-1,j]+x2old[i+1,j]+x2old[i,j-1]+x2old[i,j+1]  #cantidad de vecinos ocupados
        p=1.0-(1.0-cx2)**nvec                                     #probabilidad de ser colonizado= 1 - prob de quedar libre 
        rnd=float(g.readline())#np.random.rand()
        if (rnd < p): x2[i,j]=1                                

      ocup=0                                                       #ocupacion de herbivoros 
      if (x1old[i,j]==1 or x2old[i,j]==1): ocup=1
      if (yold[i,j]==0 and ocup==1):                               #colonizacion de y si el sitio está libre de y,y ocupado por un x 
        nvec=yold[i-1,j]+yold[i+1,j]+yold[i,j-1]+yold[i,j+1]       #cantidad de vecinos ocupados
        p=1.0-(1.0-cy)**nvec                                       #probabilidad de ser colonizado= 1 - prob de quedar libre 
        rnd=float(g.readline())#np.random.rand()
        if (rnd < p): y[i,j]=1

      #extinciones sin depredadores
      if (x1old[i,j]==1 and yold[i,j]==0):                          #extincion de x1 si el sitio esta ocupado 
        rnd=float(g.readline())#np.random.rand()
        if (rnd < ex1): x1[i,j]=0

      if (x2old[i,j]==1 and yold[i,j]==0):                          #extincion de x2 si el sitio esta ocupado 
        rnd=float(g.readline())#np.random.rand()
        if (rnd < ex2): x2[i,j]=0

      if (yold[i,j]==1):                                            #extincion de y si el sitio esta ocupado 
        rnd=float(g.readline())#np.random.rand()
        if (rnd < ey): y[i,j]=0

      #depredacion (extinción de presas donde hay depredadores)
      if (x1old[i,j]==1 and yold[i,j]==1):                          #depredacion de y sobre x1 
        rnd=float(g.readline())#np.random.rand()
        if (rnd < depx1): x1[i,j]=0

      if (x2old[i,j]==1 and yold[i,j]==1):                          #depredacion de y sobre x2 
        rnd=float(g.readline())#np.random.rand()
        if (rnd < depx2): x2[i,j]=0

      #desplazamiento por competencia (jerarquia)
      if (x1old[i,j]==1 and x2old[i,j]==1):
        rnd=float(g.readline())#np.random.rand()
        if (rnd < cx1): x2[i,j]=0
  
  fx1=sum(sum(x1))/((Lx-1)*(Ly-1))
  fx2=sum(sum(x2))/((Lx-1)*(Ly-1))
  fy =sum(sum(y)) /((Lx-1)*(Ly-1))
  f.write("	"+str(it+1)+"	"+str(fx1)+"	"+str(fx2)+"	"+str(fy)+"\n")
  
# CÁLCULOS FINAIS -----------------------------------------------------
f.close
g.close() 

File: Laguna-2015(prog-evol-2015).cs
/*
Modelo baseado em autômatos celulares
Autor:           Jhordan Silveira de Borba
E-mail:          sbjhordan@gmail.com
Site:            https://sites.google.com/view/jdan/
Modelo original: Mathematical model of livestock and wildlife: Predation and competition under environmental disturbances
Link:            https://arxiv.org/abs/1409.0024
 */

using System;               //Namespace padrão System
using System.IO;            // include the System.IO namespace
using System.Globalization; //Mudar a cultura (pontuação)


namespace Vida_selvagem_e_Pecuária
{
    class Program
    {
        static double[] Densidade(int[,] x1, int[,] x2, int[,] y, int Lx, int Ly) //Static para ser acesssível sem ser objeto
        {
            double[] f = { 0, 0, 0 };
            for (int i = 1; i < Lx - 1; i++)
            {
                for (int j = 1; j < Ly - 1; j++)
                {
                    f[0] = f[0] + x1[i, j];
                    f[1] = f[1] + x2[i, j];
                    f[2] = f[2] + y[i, j];
                }
            }
            f[0] /= (Lx - 1) * (Ly - 1);
            f[1] /= (Lx - 1) * (Ly - 1);
            f[2] /= (Lx - 1) * (Ly - 1);
            return f;
        }
        static void Main(string[] args)
        {
            CultureInfo.CurrentCulture = new CultureInfo("en-US"); //Mudamos a cultura

            //Condições iniciais
            int maxt = 15000; //Tempo total de cada realização
            int Lx = 100;   //Tamanho na dimensão X
            int Ly = 100;   //Tamanho na dimensão y

            //Frações iniciais
            double D = 0.3;
            double x10 = 0.5;
            double x20 = 0.5;
            double y0 = 0.5;

            //Taxas
            double cx1 = 0.05;
            double cx2 = 0.1;
            double cy = 0.015;

            double ex1 = 0.05;
            double ex2 = 0.01;
            double ey = 0.017;

            double mx1y = 0.2;
            double mx2y = 0.8;

            double depx1 = ex1 + mx1y;
            double depx2 = ex2 + mx2y;

            //Random rnd = new Random(); //Gerador de número aleatório alimentado pelo sistema
            StreamReader file = new System.IO.StreamReader("psd.dat");  //Método para ler arquivos      
            // Convert.ToDouble(file.ReadLine()) -> Convert.ToDouble(file.ReadLine())

            //Inicializa as matrizes
            int[,] s = new int[Lx, Ly];
            int[,] x1 = new int[Lx, Ly];
            int[,] x2 = new int[Lx, Ly];
            int[,] y = new int[Lx, Ly];

            for (int i = 1; i < Lx - 1; i++)
            {
                for (int j = 1; j < Ly - 1; j++)
                {
                    double rnd = Convert.ToDouble(file.ReadLine());
                    s[i, j] = (rnd < D) ? 0 : 1;
                    rnd = Convert.ToDouble(file.ReadLine());
                    x1[i, j] = (rnd < x10 && s[i, j] == 1) ? 1 : 0;
                    rnd = Convert.ToDouble(file.ReadLine());
                    x2[i, j] = (rnd < x20 && s[i, j] == 1) ? 1 : 0;
                    rnd = Convert.ToDouble(file.ReadLine());
                    y[i, j] = ((x1[i, j] == 1 || x2[i, j] == 1) && (rnd < y0)) ? 1 : 0;
                }
            }
            //Laço temporal
            double[] f = Program.Densidade(x1, x2, y, Lx, Ly);
            File.AppendAllText("Dados.dat", "\t" + 0 + "\t" + f[0] + "\t" + f[1] + "\t" + f[2] + "\n");
            Console.WriteLine("\t" + 0 + "\t" + f[0] + "\t" + f[1] + "\t" + f[2] + "\n");

            for (int it = 0; it < maxt; it++)
            {
                int[,] x1old = x1.Clone() as int[,];
                int[,] x2old = x2.Clone() as int[,];
                int[,] yold = y.Clone() as int[,];

                //Percorrer a malha
                for (int i = 1; i < Lx - 1; i++)
                {
                    for (int j = 1; j < Ly - 1; j++)
                    {
                        //Colonizações
                        //Guanaco
                        if (x1old[i, j] == 0 && s[i, j] == 1)
                        {
                            int nvec = x1old[i - 1, j] + x1old[i + 1, j] + x1old[i, j - 1] + x1old[i, j + 1];
                            double p = 1 - Math.Pow((1.0 - cx1), nvec);
                            double rnd = Convert.ToDouble(file.ReadLine());
                            x1[i, j] = (rnd < p) ? 1 : x1[i, j];
                        }
                        //Ovelha
                        if (x2old[i, j] == 0 && s[i, j] == 1 && x1old[i, j] == 0)
                        {
                            int nvec = x2old[i - 1, j] + x2old[i + 1, j] + x2old[i, j - 1] + x2old[i, j + 1];
                            double p = 1 - Math.Pow((1.0 - cx2), nvec);
                            double rnd = Convert.ToDouble(file.ReadLine());
                            x2[i, j] = (rnd < p) ? 1 : x2[i, j];
                        }
                        //Puma
                        int ocup = 0;
                        ocup = (x1old[i, j] == 1 || x2old[i, j] == 1) ? 1 : 0;
                        if (yold[i, j] == 0 && ocup == 1)
                        {
                            int nvec = yold[i - 1, j] + yold[i + 1, j] + yold[i, j - 1] + yold[i, j + 1];
                            double p = 1 - Math.Pow((1.0 - cy), nvec);
                            double rnd = Convert.ToDouble(file.ReadLine());
                            y[i, j] = (rnd < p) ? 1 : y[i, j];
                        }
                        //Extinção sem predador
                        //Guanaco
                        if (x1old[i, j] == 1 && yold[i, j] == 0)
                        {
                            double rnd = Convert.ToDouble(file.ReadLine());
                            x1[i, j] = (rnd < ex1) ? 0 : x1[i, j]; 
                        }
                        //Ovelha
                        if (x2old[i, j] == 1 && yold[i, j] == 0)
                        {
                            double rnd = Convert.ToDouble(file.ReadLine());
                            x2[i, j] = (rnd < ex2) ? 0 : x2[i, j];
                        }
                        //Puma
                        if (yold[i, j] == 1)
                        { 
                            double rnd = Convert.ToDouble(file.ReadLine()); 
                            y[i, j] = (rnd < ey) ? 0 : y[i, j];
                        }
                        //Depredação
                        //Guanaco
                        if (x1old[i, j] == 1 && yold[i, j] == 1)
                        { 
                            double rnd = Convert.ToDouble(file.ReadLine());
                            x1[i, j] = (rnd < depx1) ? 0 : x1[i, j]; 
                        }
                        //Ovelha
                        if (x2old[i, j] == 1 && yold[i, j] == 1)
                        { 
                            double rnd = Convert.ToDouble(file.ReadLine());
                            x2[i, j] = (rnd < depx2) ? 0 : x2[i, j]; 
                        }
                        //Deslocamento
                        if (x1old[i, j] == 1 && x2old[i, j] == 1)
                        { 
                            double rnd = Convert.ToDouble(file.ReadLine()); 
                            x2[i, j] = (rnd < cx1) ? 0 : x2[i, j]; 
                        }
                    }
                }
                f = Program.Densidade(x1, x2, y, Lx, Ly);
                File.AppendAllText("Dados.dat", "\t" + (it + 1) + "\t" + f[0] + "\t" + f[1] + "\t" + f[2] + "\n");
            }
            Console.WriteLine("\t" + 0 + "\t" + f[0] + "\t" + f[1] + "\t" + f[2] + "\n");
            file.Close();
            Console.ReadKey();
        }
    }
}